src/main/kotlin/org/rust/ide/inspections/RsNeedlessLifetimesInspection.kt

```diff
@@ -0,0 +1,240 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.openapi.util.TextRange
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.types.*
+import org.rust.lang.core.types.ty.hasReEarlyBounds
+import org.rust.stdext.chain
+
+/**
+ * Checks for lifetime annotations which can be removed by relying on lifetime elision.
+ * Corresponds to needless_lifetimes lint from Rust Clippy.
+ */
+class RsNeedlessLifetimesInspection : RsLocalInspectionTool() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): RsVisitor = object : RsVisitor() {
+        override fun visitFunction(fn: RsFunction) {
+            if (hasWhereLifetimes(fn.whereClause)) return
+
+            val boundsLifetimes = fn.typeParameters
+                .flatMap { it.bounds }
+                .map { it.bound }
+                .flatMap { bound ->
+                    val collector = LifetimesCollector()
+                    bound.accept(collector)
+                    if (collector.lifetimes.any { it is ReferenceLifetime.Named }) return
+
+                    val lifetimes = bound.traitRef?.path?.typeArgumentList?.lifetimeList.orEmpty()
+                    if (lifetimes.any { !it.isStatic && !it.isElided }) return
+                    lifetimes
+                }
+            if (couldUseElision(fn, boundsLifetimes)) {
+                holder.registerProblem(
```

Comment:
I'd extract it to `private fun registerProblem(holder: ProblemsHolder, fn: RsFunction)`